### DISCUSSION TITLE
v2_dev: ESM + TS Monorepo Cache Busting

### DISCUSSION THREAD

**Discussion by jaschaio** [remix-run/remix#7125] [Q&A]
I am following the [ES Modules + typescript example](https://github.com/xHomu/remix-v2-server/commit/ba20097d5739e676d986f8e667b53dfcd059e578) and it works fine for UI changes and as well for anything that is placed within `serverDependenciesToBundle`. But I have some other monorepo dependencies that are intentionally not bundled into the server (so coming from `node_modules` instead), included within [`watchPaths`](https://remix.run/docs/en/1.17.1/file-conventions/remix-config#watchpaths) but that don't get updated when the server reloads. The server reloads successfully, but the module versions are not updated.

It makes sense because by following the example I am only ["cache busting" the `BUILD_PATH`](https://github.com/xHomu/remix-v2-server/commit/ba20097d5739e676d986f8e667b53dfcd059e578#diff-1ba718c1eb8aa39cd20c2562d92523068c734d75f54655e97d652b992d9b4259R99) but nothing from `node_modules` – any suitable approach to cache bust as well anything specific within `node_modules`?

Maybe by implementing a custom loader on top of `ts-node`?

**Comment by jaschaio** [comment]
This approach using a [custom node `--loader` works](https://nodejs.org/api/esm.html#esm_loaders) for me because apparently node uses whatever is returned by resolve as the cache key, but it creates a memory leak because previously imported modules are never cleared:

```js
export const resolve = async ( specifier, parentModuleURL, defaultResolve ) => {

    const { url, format } = await defaultResolve( specifier, parentModuleURL );

    if ( ! url.includes( 'packages/services' ) || ! process.env.ESM_CACHE )
        return { url, format };

    const versioned = new URL( url );

    versioned.searchParams.set( 'v', process.env.ESM_CACHE );

    return { url: versioned.toString(), format };

};
```

```diff
// Within createDevRequestHandler
+    if ( ! process.env.ESM_CACHE )
+        process.env.ESM_CACHE = new Date().getTime();
 
    const handleServerUpdate = async () => {

+        process.env.ESM_CACHE = new Date().getTime();

        // 1. re-import the server build
        build = await reimportServer( BUILD_PATH );

        // 2. tell dev server that this app server is now up-to-date and ready
        broadcastDevReady( build );

    };

// ...
```

Any better suggestions?

**Reply by pcattori** [reply]
Can you setup your monorepo to rebuild your other packages when they change and then add the `node_modules/` paths that correspond to those packages to your `watchPaths`?

**Comment by MichaelDeBoey** [comment]
CC/ @pcattori
